Feature/764 state migrator

zzeekk opened the pull request:
see #764

pgruetter commented on sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRunState.scala:
```diff
@@ -125,11 +133,47 @@ private[smartdatalake] object ActionDAGRunState {
+  private val stateMigrators: Seq[StateMigratorDef] = Seq(
+    new StateMigratorDef3To4()
+  ).sortBy(_.versionFrom) // force ordering
+  assert(stateMigrators.groupBy(_.versionFrom).forall(_._2.size == 1)) // check that versionFrom is unigue
```
```suggestion
  assert(stateMigrators.groupBy(_.versionFrom).forall(_._2.size == 1)) // check that versionFrom is unique
```

pgruetter commented on sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala:
```diff
@@ -0,0 +1,81 @@
+ * Migrate state from format version 3 to 4:
+ * - the structure of `actionState.results` was cleaned up.
+ *   instead of having `results.subFeed`, the attributes of `subFeed` have been moved directly into `result`.
+ * - Additionally `actionState.inputIds[].id` was cleaned up as `inputIds[]`, and the the same applies for `actionState.outputIds`.
```
```suggestion
 * - Additionally `actionState.inputIds[].id` was cleaned up as `inputIds[]`, and the same applies for `actionState.outputIds`.
```

pgruetter commented on sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala:
```diff
@@ -0,0 +1,81 @@
+  override val versionFrom = 3
+  override val versionTo = 4
+  override def migrate(json: JValue): JValue = {
+    assert(json \ "runStateFormatVersion" match { case JInt(version) => version <= versionFrom}, s"Version should be equals or less than $versionFrom")
```
Shouldn't we assume that version = versionFrom ? The upgrade procedure is probably strictly working for one version only.

zzeekk replied:
Then we could not migrate from version 1/2 to version 4. As there are no changes needed to migrate from version 1to2 or 2to3, no StateMigratorDef exists for these steps.

pgruetter replied:
OK, makes sense. Thanks!

pgruetter commented on sdl-lang/src/main/scala/io/smartdatalake/meta/state/StateMigrator.scala:
```diff
@@ -0,0 +1,86 @@
+import org.json4s.jackson.{JsonMethods, prettyJson}
+import scopt.OptionParser
+
+case class StateMigratorConfig(statePath: Option[String] = None)
```
Can you please add an overall ScalaDoc ?

zzeekk replied:
Yes, sure. Done.

pgruetter reviewed (commented):
Great feature, just added a few comments. 
The assumption should be, that we strictly update one version to the next IMHO.

iliassoto commented:
I did some testing for this feature. The mentioned files are attached:

1. I ran "myApp.conf" and with SDLB 2.3.0., resulting in the state "myAppName.1.1.json"
2. I then created a new project with the SDLB built from this branch, and tried to run the pipeline again using the old state file, where an error was thrown (see file). 

The JSON itself is able to be parsed: 

```scala
  // read state from json
def fromJson(stateJson: String): ActionDAGRunState = {
      val jValue = JsonMethods.parse(stateJson) //WORKS
      val migratedJValue = checkStateFormatVersionAndMigrate(jValue).getOrElse(jValue) //ERROR HAPPENS HERE
      // extract into class structures
      migratedJValue.extract[ActionDAGRunState]
    } catch {
      case ex: Exception => throw new IllegalStateException(s"Unable to parse state from json: ${ex.getMessage}", ex)
    }
  }
```

[error.log](https://github.com/smart-data-lake/smart-data-lake/files/14001698/error.log)
[myApp.hocon.txt](https://github.com/smart-data-lake/smart-data-lake/files/14001699/myApp.hocon.txt)
[myAppName.1.1.json](https://github.com/smart-data-lake/smart-data-lake/files/14001700/myAppName.1.1.json)

The operation _json \ "runStateFormatVersion"_ results in a _JNothing_-Type, which throws the error during the assert.

```scala
  override def migrate(json: JValue): JValue = {
    assert(json \ "runStateFormatVersion" match { case JInt(version) => version <= versionFrom}, s"Version should be equals or less than $versionFrom")
    json.transformField {
      case (name, actionsState) if name == "actionsState" =>
        (name, actionsState.transformField {
          case (name, results) if name == "results" =>
            (name, JArray(results.children.map {
              result =>
```

I guess we can just add another case that also handles JNothing for versions that don't have that node in their JSON files?

zzeekk commented:
> ```scala
> I guess we can just add another case that also handles JNothing for versions that don't have that node in their JSON files?
> ```
You're right, early state files without version attribute produce an error. I added the JNothing case, and fixed adding the version attribute afterwards.
